src/macos/icons.rs

```diff
@@ -0,0 +1,79 @@
+use anyhow::{bail, Result};
+use cocoa::base::id;
+use objc::{class, msg_send, sel, sel_impl};
+use once_cell::sync::Lazy;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::path::Path;
+use std::path::PathBuf;
+use std::{collections::hash_map::Entry, sync::Mutex};
+use sysinfo::{PidExt, ProcessExt, ProcessRefreshKind, System, SystemExt};
+
+pub static ICON_CACHE: Lazy<Mutex<IconCache>> = Lazy::new(|| Mutex::new(IconCache::default()));
+
+#[derive(Default)]
+pub struct IconCache {
+    /// executable name -> icon hash
+    executables: HashMap<PathBuf, u64>,
+    /// icon hash -> png bytes
+    icons: HashMap<u64, Vec<u8>>,
+}
+
+impl IconCache {
+    pub fn get_png(&mut self, executable: PathBuf) -> Result<&Vec<u8>> {
+        match self.executables.entry(executable) {
+            Entry::Occupied(e) => {
+                // Guaranteed to exist because we never clear the cache.
+                Ok(self.icons.get(e.get()).unwrap())
+            }
+            Entry::Vacant(e) => {
+                let icon = unsafe { png_data_for_executable(e.key())? };
+                let mut hasher = DefaultHasher::new();
+                icon.hash(&mut hasher);
+                let icon_hash = hasher.finish();
+                e.insert(icon_hash);
+                let icon = self.icons.entry(icon_hash).or_insert(icon);
+                Ok(icon)
+            }
+        }
+    }
+}
+
+unsafe fn png_data_for_executable(executable: &Path) -> Result<Vec<u8>> {
+    let mut sys = System::new();
+    sys.refresh_processes_specifics(ProcessRefreshKind::new());
```

Comment:
Potentially dumb question: Can we use https://developer.apple.com/documentation/appkit/nsworkspace/1528158-icon instead of enumerating processes?

Reply:
I think that method works for files such as for example images, pdf etc etc, but for sure I give it a try!

Reply:
(from Slack: @emanuele-em is right, doesn't work :))